src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -82,6 +86,20 @@
 
     public static final String UPDATE_RESTORE_ACTION_NAME = "internal:cluster/snapshot/update_restore";
 
+    private static final ImmutableSet<String> UNMODIFIABLE_SETTINGS = ImmutableSet.of(
+            SETTING_NUMBER_OF_SHARDS,
+            SETTING_VERSION_CREATED,
+            SETTING_LEGACY_ROUTING_HASH_FUNCTION,
+            SETTING_LEGACY_ROUTING_USE_TYPE);
+
+    // It's OK to change some settings, but we shouldn't allow simply removing them
+    private static final ImmutableSet<String> UNREMOVABLE_SETTINGS = ImmutableSet.<String>builder()
+            .addAll(UNMODIFIABLE_SETTINGS)
+            .add(SETTING_NUMBER_OF_REPLICAS)
```

Comment:
Defaults for the number of replicas are [non-trivial](https://github.com/elasticsearch/elasticsearch/blob/master/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java#L335). Not sure I want to replicate this logic here especially considering that it might change over time. I should probably add `SETTING_AUTO_EXPAND_REPLICAS` to the list of unremovable settings as well though.